### PR TITLE
feat: amazonsagemaker AI provider

### LINKED ISSUE #730: [Feature]: AmazonSageMaker AI provider
### Checklist

- [X] I've searched for similar issues and couldn't find anything matching
- [X] I've discussed this feature request in the [K8sGPT Slack](https://k8sgpt.slack.com) and got positive feedback

### Is this feature request related to a problem?

None

### Problem Description

_No response_

### Solution Description

k8sgpt should support SageMaker provider as well.

### Benefits

You can create your own AWS SageMaker with AI model of your choice, and use it for k8sgpt.

### Potential Drawbacks

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by dkuroczk**
Closes #730

## 📑 Description

- Adds Amazon SageMaker as a backend
- adds `endpointname` flag to `k8sgpt auth`
- `endpointname` and `providerRegion` are required for amazonsagemaker backend 
- Updated README.md

## ✅ Checks

- [x] My pull request adheres to the code style of this project
- [x] My code requires changes to the documentation
- [x] I have updated the documentation as required
- [ ] All the tests have passed

## ℹ Additional Information

Co-authored-by: Mateusz Zaremba

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -398,6 +398,58 @@ k8sgpt analyze -e -b amazonbedrock
...
+* AWS region where your SageMaker instance is created
+
+To add amazonsagemaker as a backend run:
+
+```bash
+k8sgpt auth add --backend amazonsagemaker --providerRegion eu-west-1 --endpointname endpoint-xxxxxxxxxx
+```
+
+***Note**:
+TODO: Currently access key will be requested in the CLI, you can enter anything into this.
```
Comment (AlexsJones) [inline README.md:426]: We should make this selective for the project. I'll raise an issue!

File: pkg/ai/amazonsagemaker.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+	client      *sagemakerruntime.SageMakerRuntime
+	language    string
+	model       string
+	temperature float32
+	endpoint    string
+}
+
+const (
+	// SageMaker completion parameters
+	maxNewTokens = 256
```
Comment (AlexsJones) [inline pkg/ai/amazonsagemaker.go:44]: That's a very small token count, probably not large enough for an event, Would we be able to set this something higher more like 1024?

File: pkg/ai/amazonsagemaker.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	top_P        = 0.9
```
Comment (AlexsJones) [inline pkg/ai/amazonsagemaker.go:45]: This would really cause many hallucinations - I suggest for credibility we lower this. Perhaps even parameterise it?

Comment (AlexsJones) [reply pkg/ai/amazonsagemaker.go:45]: An example of using the passed in temperature https://github.com/k8sgpt-ai/k8sgpt/blob/main/pkg/ai/azureopenai.go#L46

File: pkg/ai/amazonsagemaker.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+type Generation struct {
+	Role    string `json:"role"`
+	Content string `json:"content"`
+}
+
+func (c *SageMakerAIClient) Configure(config IAIConfig, language string) error {
+
+	// Create a new AWS session
+	sess := session.Must(session.NewSessionWithOptions(session.Options{
+		Config:            aws.Config{Region: aws.String(config.GetProviderRegion())},
```
Comment (AlexsJones) [inline pkg/ai/amazonsagemaker.go:57]: I would make it clear in the README that we use the default region provided in the envs

File: pkg/ai/amazonsagemaker.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+func (c *SageMakerAIClient) GetCompletion(ctx context.Context, prompt string, promptTmpl string) (string, error) {
+	// Create a completion request
+
+	if len(promptTmpl) == 0 {
+		promptTmpl = PromptMap["default"]
+	}
+	// TODO: extract all paramseters to config
+	data := map[string]interface{}{
```
Comment (AlexsJones) [inline pkg/ai/amazonsagemaker.go:77]: If you have the raw response you can use https://transform.tools/json-to-go

File: pkg/ai/amazonsagemaker.go [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+			"max_new_tokens": maxNewTokens,
+			"top_p":          top_P,
+			"temperature":    c.temperature,
+		},
+	}
+	// Convert data to []byte
+	bytesData, err := json.Marshal(data)
+	if err != nil {
+		fmt.Println("Error:", err)
+		log.Fatal(err)
```
Comment (AlexsJones) [inline pkg/ai/amazonsagemaker.go:100]: If you log fatal here, this won't return

File: pkg/ai/amazonsagemaker.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+	}
+
+	// Define a slice of Generations
+	var generations []struct {
+		Generation Generation `json:"generation"`
+	}
+
+	err = json.Unmarshal([]byte(string(result.Body)), &generations)
+	if err != nil {
+		log.Fatal(err)
```
Comment (AlexsJones) [inline pkg/ai/amazonsagemaker.go:127]: Advise to remove fatal here

File: pkg/ai/amazonsagemaker.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+		return "", err
+	}
+
+	// Access the content
+	content := generations[0].Generation.Content
```
Comment (AlexsJones) [inline pkg/ai/amazonsagemaker.go:132]: length check here please!

File: pkg/ai/amazonsagemaker.go [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+	if err != nil {
+		color.Red("error getting completion: %v", err)
+		return "", err
+	}
+
+	err = cache.Store(cacheKey, base64.StdEncoding.EncodeToString([]byte(response)))
+
+	if err != nil {
+		color.Red("error storing value to cache: %v", err)
+		return "", nil
```
Comment (AlexsJones) [inline pkg/ai/amazonsagemaker.go:153]: return the error please!

**Review by AlexsJones — CHANGES_REQUESTED** [review]

File: pkg/ai/amazonsagemaker.go (hunk 2 above)
Comment (dkuroczk) [reply pkg/ai/amazonsagemaker.go]: It's parametrized now and default is changed to 2048

File: pkg/ai/amazonsagemaker.go (hunk 4 above)
Comment (dkuroczk) [reply pkg/ai/amazonsagemaker.go:71]: k8sgpt will use ProviderRegion set in config, not one from aws cli config or env, I added that info in README
https://github.com/dkuroczk/k8sgpt/blob/feat/add-amazonsagemaker/README.md?plain=1#L418

File: README.md (hunk 1 above)
Comment (dkuroczk) [reply README.md]: I created a list of passwordlessProviders

File: pkg/ai/amazonsagemaker.go [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+		EndpointName:     aws.String(c.endpoint),
+		ContentType:      aws.String("application/json"), // Set the content type as per your model's requirements
+		Accept:           aws.String("application/json"), // Set the accept type as per your model's requirements
+		CustomAttributes: aws.String("accept_eula=true"),
+	}
+
+	// Call the InvokeEndpoint function
+	result, err := c.client.InvokeEndpoint(input)
+	if err != nil {
+		log.Fatal(err)
```
Comment (AlexsJones) [inline pkg/ai/amazonsagemaker.go:126]: This won't block the review, but please remove the fatal here

**Review by AlexsJones — APPROVED** [review]

File: pkg/ai/amazonsagemaker.go (hunk 10 above)
Comment (dkuroczk) [reply pkg/ai/amazonsagemaker.go:126]: I don't know how I missed that, Thanks!

**Pull request merged**